pkg/util/k8s/transform.go

```diff
@@ -0,0 +1,83 @@
+// Copyright 2024 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package k8s
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/client-go/tools/cache"
+)
+
+// NewTrimmer returns a cache.TransformFunc that can be used to trim objects stored in informers.
+// The function must be idempotent before client-go v0.31.0 to avoid a race condition happening when objects were
+// accessed during Resync operation, see https://github.com/kubernetes/kubernetes/issues/124337.
+// But it's generally more efficient to avoid trimming the same object more than once.
+func NewTrimmer(extraTrimmers ...cache.TransformFunc) cache.TransformFunc {
+	return func(obj interface{}) (interface{}, error) {
+		accessor, err := meta.Accessor(obj)
+		if err != nil {
+			return obj, nil
+		}
+		// It means the objects has been trimmed.
+		if accessor.GetManagedFields() == nil {
+			return obj, nil
+		}
+		// Trim common fields for all objects.
+		accessor.SetManagedFields(nil)
```

Comment:
So just to confirm, we don't have the same issue for `managedFields` as for `ownerReference`?
Is it because of this:

> Note that setting the managedFields to an empty list will not reset the field. This is on purpose, so managedFields never get stripped by clients not aware of the field.

From https://kubernetes.io/docs/reference/using-api/server-side-apply/#clearing-managedfields

If yes, maybe we could add a comment explaining why this is safe?

Reply:
sure, added a few notes and explanations.